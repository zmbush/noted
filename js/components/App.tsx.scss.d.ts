export const grow: string;
export const root: string;
