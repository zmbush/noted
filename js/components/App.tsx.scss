// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//

.root {
  width: '100%';
  @media print {
    overflow: visible !important;
    column-count: 2;
    column-width: 200;
  }
}

.grow {
  flex-grow: 1;
}
