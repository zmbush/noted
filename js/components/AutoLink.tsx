// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import * as React from 'react';
import { Link } from 'react-router-dom';
import { LinkIdMap } from 'data/selectors';

type LinkProps = {
  text: string;
  ids: Set<number>;
};

const LinkedText = ({ ids, text }: LinkProps) => {
  if (ids.size === 1) {
    const id = ids.values().next().value;
    return <Link to={`/note/${id}`}>{text}</Link>;
  }
  if (ids.size > 1) {
    const theseIds = Array.from(ids.values()).join(',');
    return <Link to={`/disambiguation/${theseIds}`}>{text}</Link>;
  }
  return <>text</>;
};

type Props = {
  titles: LinkIdMap;
  children: string;
};

export default class AutoLink extends React.Component<Props> {
  render() {
    const { children, titles } = this.props;
    let body: any[] = [children];
    let keyIx = 1;

    titles.forEach((value, key) => {
      let newBody: any[] = [];
      body.forEach(part => {
        if (typeof part === 'string' || part instanceof String) {
          const elements = part.split(new RegExp(key, 'i')).reduce((r, a, ix, arr) => {
            r.push(a);
            if (ix + 1 < arr.length) {
              r.push(<LinkedText key={`linked-text-${keyIx}`} text={key} ids={value} />);
              keyIx += 1;
            }
            return r;
          }, []);
          newBody = newBody.concat(elements);
        } else {
          newBody.push(part);
        }
      });
      body = newBody;
    });

    return body;
  }
}
