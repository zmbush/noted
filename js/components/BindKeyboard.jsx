// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import React from 'react';
import Mousetrap from 'mousetrap';
import PropTypes from 'prop-types';

export default class BindKeyboard extends React.Component {
  static propTypes = {
    keys: PropTypes.string.isRequired,
    action: PropTypes.string,
    callback: PropTypes.func.isRequired,
    global: PropTypes.bool,
  };

  constructor(props) {
    super(props);

    this.main = React.createRef();
  }

  componentDidMount() {
    if (this.props.children) {
      this.mousetrap = new Mousetrap(this.main.current);
    } else {
      this.mousetrap = Mousetrap;
    }

    this.mousetrap.bind(
      this.props.keys,
      this.props.callback,
      this.props.action
    );
  }

  componentWillUnmount() {
    this.mousetrap.unbind(this.props.keys, this.props.action);
  }

  render() {
    if (this.props.children) {
      return <div ref={this.main}>{this.props.children}</div>;
    } else {
      return null;
    }
  }
}
