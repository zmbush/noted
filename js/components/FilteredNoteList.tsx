// Copyright 2018 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import * as React from 'react';
import { useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';

import NoteList from 'components/NoteList';
import { getNoteEntities } from 'data/selectors';

interface Props {
  search: string;
  depth: number;
}

const FilteredNoteList = ({ search, depth }: Props) => {
  const { ids = '' } = useParams<'ids'>();
  const parsedIds = new Set(ids.split(',').map((i) => parseInt(i, 10)));
  const notes = useSelector(getNoteEntities);

  return (
    <NoteList
      parent_note_id={null}
      renderOnly={parsedIds}
      notes={notes}
      search={search}
      depth={depth}
    />
  );
};

export default FilteredNoteList;
