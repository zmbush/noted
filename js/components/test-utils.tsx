// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';

import * as React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import { createStore, store as defaultStore } from 'data/store';

type ComponentToRender = Parameters<typeof render>[0];
type Config = {
  route?: string;
  store?: typeof defaultStore;
  history?: ReturnType<typeof createMemoryHistory>;
} & Omit<Parameters<typeof render>[1], 'wrapper'>;
const customRender = (
  ui: ComponentToRender,
  { route = '/', store = createStore(), ...renderOptions }: Config = {},
) => {
  const Wrapper = ({ children }: any) => (
    <Provider store={store}>
      <MemoryRouter initialEntries={[route]}>{children}</MemoryRouter>
    </Provider>
  );

  return {
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
    store,
  };
};

export * from '@testing-library/react';
export { customRender as render };
