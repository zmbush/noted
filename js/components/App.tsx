// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
import axios from 'axios';
import debounce from 'debounce-promise';
import Mousetrap from 'mousetrap';
import { Dispatch } from 'redux';

import * as React from 'react';
import { connect } from 'react-redux';
import { Route, Routes, useNavigate } from 'react-router-dom';

import AccountCircle from '@mui/icons-material/AccountCircle';
import HomeIcon from '@mui/icons-material/Home';
import MenuIcon from '@mui/icons-material/Menu';
import SearchIcon from '@mui/icons-material/Search';
import AppBar from '@mui/material/AppBar';
import Grid from '@mui/material/Grid';
import IconButton from '@mui/material/IconButton';
import InputBase from '@mui/material/InputBase';
import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import { Theme, alpha } from '@mui/material/styles';
import { WithStyles } from '@mui/styles';
import createStyles from '@mui/styles/createStyles';
import withStyles from '@mui/styles/withStyles';

import BindKeyboard from 'components/BindKeyboard';
import FilteredNoteList from 'components/FilteredNoteList';
import LogIn from 'components/LogIn';
import Note, { InnerNote } from 'components/Note';
import NoteList from 'components/NoteList';
import { updateNote as updateNoteAction, deleteNote, logOut } from 'data/actions';
import { getTopLevelNotes } from 'data/selectors';
import { NoteData, AppState } from 'data/types';

const styles = (theme: Theme) =>
  createStyles({
    root: {
      width: '100%',
      '@media print': {
        overflow: 'visible !important',
        columnCount: 2,
        columnWidth: 200,
        // columnBreakInside: 'avoid',
      },
    },
    grow: {
      flexGrow: 1,
    },
    menuButton: {
      marginLeft: -12,
      marginRight: 20,
    },
    contentRoot: {
      '@media print': {
        marginTop: 0,
        display: 'block',
      },
      marginTop: 75,
    },
    title: {
      display: 'none',
      // [theme.breakpoints.up('sm')]: {
      //   display: 'block',
      // },
    },
    search: {
      position: 'relative',
      borderRadius: theme.shape.borderRadius,
      backgroundColor: alpha(theme.palette.common.white, 0.15),
      '&:hover': {
        backgroundColor: alpha(theme.palette.common.white, 0.25),
      },
      marginLeft: 0,
      marginRight: theme.spacing(1),
      width: '100%',
      // [theme.breakpoints.up('sm')]: {
      //   marginLeft: theme.spacing(1),
      //   width: 'auto',
      // },
    },
    searchIcon: {
      width: theme.spacing(9),
      height: '100%',
      position: 'absolute',
      pointerEvents: 'none',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
    },
    inputRoot: {
      color: 'inherit',
      width: '100%',
    },
    inputInput: {
      paddingTop: theme.spacing(1),
      paddingRight: theme.spacing(1),
      paddingBottom: theme.spacing(1),
      paddingLeft: theme.spacing(10),
      transition: theme.transitions.create('width'),
      width: '100%',
      // [theme.breakpoints.up('sm')]: {
      //   width: 120,
      //   '&:focus': {
      //     width: 200,
      //   },
      // },
    },
    newButton: {
      margin: theme.spacing(1),
    },
    leftIcon: {
      marginRight: theme.spacing(1),
    },
    iconSmall: {
      fontSize: 20,
    },
    noPrint: {
      '@media print': {
        display: 'none',
      },
    },
  });

interface Props extends WithStyles<typeof styles> {
  notes: Map<number, NoteData>;
  is_signed_in: boolean;
  updateNote: (note: NoteData) => void;
  deleteNote: (id: number) => void;
  logOut: () => void;
}

const App = ({
  classes,
  deleteNote: doDeleteNote,
  updateNote: doUpdateNote,
  logOut: doLogOut,
  is_signed_in: isSignedIn,
  notes,
}: Props) => {
  const searchInput = React.useRef<HTMLInputElement>();
  const firstNote = React.useRef<InnerNote>();
  const [userMenuEl, setUserMenuEl] = React.useState<HTMLElement>(null);
  const [searchInputValue, setSearchInputValue] = React.useState('');
  const [newNote, setNewNote] = React.useState(false);
  const [search, setSearch] = React.useState('');
  const [debouncedSearch, _] = React.useState<(v: string) => Promise<string>>(() =>
    debounce(async (v) => v, 100),
  );
  const navigate = useNavigate();
  React.useEffect(() => {
    document.title = `noted`;
  }, []);

  const isUserMenuOpen = Boolean(userMenuEl);

  const startSearch = (e: Event) => {
    e.preventDefault();
    searchInput.current.focus();
  };

  const create = () => {
    setNewNote(true);
  };

  const createNew = (e: React.SyntheticEvent) => {
    e.preventDefault();
    create();
  };

  const startEdit = (e: React.SyntheticEvent) => {
    e.preventDefault();
    if (firstNote.current) {
      // TODO: Fix this.
      // firstNote.current.startEdit();
      createNew(e);
    } else {
      createNew(e);
    }
  };

  const doSearch = async (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchInputValue(e.target.value);
    setSearch(await debouncedSearch(e.target.value));
  };

  const updateNote = (note?: NoteData) => {
    if (note) {
      doUpdateNote(note);
    }
    setNewNote(false);
  };

  const cancelSearch = (e: Mousetrap.ExtendedKeyboardEvent, _combo?: string) => {
    e.preventDefault();
    setSearchInputValue('');
    setSearch('');
  };

  const createNewShortcut = (
    e: Mousetrap.ExtendedKeyboardEvent | React.SyntheticEvent,
    _combo?: string,
  ) => {
    e.preventDefault();
    create();
  };

  const closeUserMenu = () => {
    setUserMenuEl(null);
  };

  const openUserMenu = (e: React.MouseEvent<HTMLElement>) => {
    setUserMenuEl(e.currentTarget);
  };

  const signOut = async (_e: React.SyntheticEvent) => {
    setUserMenuEl(null);
    await axios.post('/api/sign_out');
    doLogOut();
  };

  const filteredNoteList = (
    <FilteredNoteList
      depth={1}
      search={search}
      updateNote={updateNote}
      deleteNote={doDeleteNote}
      firstNoteRef={firstNote}
    />
  );

  return (
    <div className={classes.root}>
      <AppBar className={classes.noPrint}>
        <Toolbar>
          <Routes>
            <Route
              path='/'
              element={
                <IconButton
                  className={classes.menuButton}
                  aria-label='Menu'
                  color='inherit'
                  size='large'
                >
                  <MenuIcon />
                </IconButton>
              }
            />
            <Route
              path='/*'
              element={
                <IconButton
                  className={classes.menuButton}
                  aria-label='Menu'
                  color='inherit'
                  onClick={() => {
                    navigate('/');
                  }}
                  size='large'
                >
                  <HomeIcon />
                </IconButton>
              }
            />
          </Routes>
          <Typography className={classes.title} variant='h6' color='inherit' noWrap>
            Noted
          </Typography>
          <div className={classes.grow} />
          <BindKeyboard keys='esc' callback={cancelSearch}>
            <BindKeyboard keys='ctrl+o' callback={createNewShortcut}>
              <div className={classes.search}>
                <div className={classes.searchIcon}>
                  <SearchIcon />
                </div>
                <form onSubmit={startEdit}>
                  <InputBase
                    inputProps={{
                      ref: searchInput,
                    }}
                    placeholder='Search...'
                    classes={{
                      root: classes.inputRoot,
                      input: classes.inputInput,
                    }}
                    value={searchInputValue}
                    onChange={doSearch}
                  />
                </form>
              </div>
            </BindKeyboard>
          </BindKeyboard>
          <IconButton aria-haspopup='true' onClick={openUserMenu} color='inherit' size='large'>
            <AccountCircle />
          </IconButton>
        </Toolbar>
      </AppBar>
      <Grid container spacing={2} className={classes.contentRoot}>
        {newNote ? (
          <Grid item xs={12}>
            <Note
              new
              depth={1}
              search={search}
              note={{
                id: -1,
                title: search,
                tags: [],
                body: '',
                created_at: '',
                updated_at: '',
                user_id: 0,
              }}
              updateNote={updateNote}
              deleteNote={deleteNote}
            />
          </Grid>
        ) : null}
        <Routes>
          <Route
            path='/'
            element={
              <NoteList
                createFromSearch={createNewShortcut}
                parent_note_id={null}
                depth={1}
                notes={notes}
                search={search}
                updateNote={updateNote}
                deleteNote={doDeleteNote}
                firstNoteRef={firstNote}
              />
            }
          />
          <Route path='/note/:ids' element={filteredNoteList} />
          <Route path='/disambiguation/:ids' element={filteredNoteList} />
        </Routes>
      </Grid>

      <BindKeyboard keys='/' callback={startSearch} />
      <LogIn open={!isSignedIn} />
      <Menu
        anchorEl={userMenuEl}
        open={isUserMenuOpen}
        onClose={closeUserMenu}
        anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
      >
        <MenuItem onClick={signOut}>
          <p>Sign Out</p>
        </MenuItem>
      </Menu>
    </div>
  );
};

const mapStateToProps = (state: AppState) => ({
  notes: getTopLevelNotes(state),
  is_signed_in: state.user.is_signed_in,
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  updateNote(data: NoteData) {
    dispatch(updateNoteAction(data));
  },

  deleteNote(id: number) {
    dispatch(deleteNote(id));
  },

  logOut() {
    dispatch(logOut());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(App));
