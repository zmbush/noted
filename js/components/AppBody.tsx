// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import Mousetrap from 'mousetrap';

import * as React from 'react';
import { Route, Routes } from 'react-router-dom';

import { Grid } from '@mui/material';

import FilteredNoteList from 'components/FilteredNoteList';
import Note from 'components/Note';
import NoteList from 'components/NoteList';
import { NoteWithTags } from 'data/types';

type NewNoteProps = {
  newNote: boolean;
  search: string;
  onNewNoteCancel: () => void;
};

export const NewNote = ({ newNote, search, onNewNoteCancel }: NewNoteProps) => {
  if (!newNote) {
    return null;
  }
  return (
    <Grid item xs={12}>
      <Note
        depth={1}
        search={search}
        note={{
          title: search,
          body: '',
        }}
        onNewNoteCancel={onNewNoteCancel}
      />
    </Grid>
  );
};

type Props = {
  notes: Map<number, NoteWithTags>;
  createNewShortcut: (
    e: Mousetrap.ExtendedKeyboardEvent | React.SyntheticEvent,
    combo?: string,
  ) => void;
} & NewNoteProps;

const AppBody = ({ notes, createNewShortcut, newNote, search, onNewNoteCancel }: Props) => {
  const filteredNoteList = <FilteredNoteList depth={1} search={search} />;

  const mainNoteList = (
    <NoteList
      createFromSearch={createNewShortcut}
      parent_note_id={null}
      depth={1}
      notes={notes}
      search={search}
    />
  );

  return (
    <Grid
      container
      spacing={1}
      sx={{
        '@media print': {
          marginTop: 0,
          display: 'block',
        },
        marginTop: '75px',
      }}
    >
      <NewNote newNote={newNote} search={search} onNewNoteCancel={onNewNoteCancel} />
      <Routes>
        <Route path='/' element={mainNoteList} />
        <Route path='/archive' element={mainNoteList} />
        <Route path='/note/:ids' element={filteredNoteList} />
        <Route path='/disambiguation/:ids' element={filteredNoteList} />
      </Routes>
    </Grid>
  );
};

export default AppBody;
