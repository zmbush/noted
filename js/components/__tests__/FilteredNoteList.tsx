// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import * as React from 'react';
import { Route, Routes } from 'react-router-dom';

import { render } from 'components/test-utils';
import { signInUser } from 'data/user/api';

import FilteredNoteList from '../FilteredNoteList';

describe('<FilteredNoteList />', () => {
  test('matches snapshot', async () => {
    const { store, container, findByText, history } = render(
      <Routes>
        <Route path='/note/:ids' element={<FilteredNoteList depth={1} search='' />} />
      </Routes>,
      {
        route: '/note/1',
      },
    );

    expect(container).toMatchInlineSnapshot(`<div />`);

    await store.dispatch(signInUser({ email: 'test@test.com', password: 'pass' }));

    // Should be showing note 1 by default.
    const note1 = store.getState().notes.entities[1];
    await findByText(note1.title);
    await findByText(note1.body);

    // Navigate to /note/2
    history.replace('/note/2');
    const note2 = store.getState().notes.entities[2];
    await findByText(note2.title);
    await findByText(note2.body);
  });
});
