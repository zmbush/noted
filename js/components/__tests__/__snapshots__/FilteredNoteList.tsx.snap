// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilteredNoteList /> matches snapshot 1`] = `
<withRouter(Connect(WithStyles(NoteList)))
  deleteNote={[Function]}
  depth={1}
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "",
      "pathname": "/",
      "search": "",
      "state": undefined,
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {
        "ids": "1",
      },
      "path": "",
      "url": "",
    }
  }
  notes={Map {}}
  parent_note_id={null}
  renderOnly={
    Set {
      1,
    }
  }
  search=""
  updateNote={[Function]}
/>
`;
